src/actors/common/cache.rs

```diff
@@ -167,138 +172,138 @@ impl<T: CacheItemRequest> Cacher<T> {
+                log::trace!("Creating {} at path {:?}", name, cache_path);
             }
 
-            // A file was not found. If this spikes, it's possible that the filesystem cache
-            // just got pruned.
-            metric!(counter(&format!("caches.{}.file.miss", name)) += 1);
+            let byteview = ByteView::open(temp_file.path())?;
 
-            let cache_path = get_scope_path(slf.config.cache_dir(), &key.scope, &key.cache_key);
+            metric!(
+                counter(&format!("caches.{}.file.write", name)) += 1,
+                "status" => status.as_ref(),
+            );
+            metric!(
+                time_raw(&format!("caches.{}.file.size", name)) = byteview.len() as u64,
+                "hit" => "false"
+            );
 
-            let temp_file = if let Some(ref path) = cache_path {
-                let dir = path.parent().unwrap();
-                tryf!(fs::create_dir_all(dir));
-                tryf!(NamedTempFile::new_in(dir))
-            } else {
-                tryf!(NamedTempFile::new())
+            let path = match cache_path {
+                Some(ref cache_path) => {
+                    status.persist_item(cache_path, temp_file)?;
+                    CachePath::Cached(cache_path.to_path_buf())
+                }
+                None => CachePath::Temp(temp_file.into_temp_path()),
             };
 
-            let future = request.compute(temp_file.path()).and_then(move |status| {
-                if let Some(ref cache_path) = cache_path {
-                    configure_scope(|scope| {
-                        scope.set_extra(
-                            &format!("cache.{}.cache_path", name),
-                            format!("{:?}", cache_path).into(),
-                        );
-                    });
+            Ok(request.load(key.scope.clone(), status, byteview, path))
+        });
+
+        Box::new(future)
+    }
 
-                    log::trace!("Creating {} at path {:?}", name, cache_path);
-                }
+    /// Creates a shareable channel that computes an item.
+    fn create_channel(&self, request: T, key: CacheKey) -> ComputationChannel<T::Item, T::Error> {
+        let (sender, receiver) = oneshot::channel();
 
-                let byteview = tryf!(ByteView::open(temp_file.path()));
+        let slf = self.clone();
+        let current_computations = self.current_computations.clone();
+        let remove_computation_token = CallOnDrop::new(clone!(key, || {
+            current_computations.lock().remove(&key);
+        }));
 
-                metric!(
-                    counter(&format!("caches.{}.file.write", name)) += 1,
-                    "status" => status.as_ref(),
-                );
+        // Run the computation and wrap the result in Arcs to make them clonable.
+        let channel = future::lazy(move || slf.compute(request, key))
+            .then(move |result| {
+                drop(remove_computation_token);
```

Comment:
might be a good idea to also attach your comment here about the order. What is the expectation when this is dropped for other reasons?

Reply:
Added, thanks!